raft-tech/df-arcadedb
# WIP OPA ACCM/Data access enforcement

# Overview

Arcade integration with opa. We are adding logic to be able to pull policy from opa. In turn arcade will use rendered policy to:
- make decisions what data to show to the user, ingest data, update data and delete data.

## Related Issues

> _Reference any issue(s) this PR should either close or is related to in some way._

* Closes raft-tech/data-fabric#3508

# Test Procedure

## Setup

From a new shell session in any clean/empty directory, perform the following 👇.

### Environment

1. Export your GitHub credentials to pull from the GitHub container registry.
    ```shell
    export GHP_USERNAME=your_github_username_goes_here
    export GHP_SECRET=your_github_pat_goes_here
    ```

1. Create a subdirectory for this PR test.
    ```shell
    export PR_HOME=$(pwd)/pr-test
    mkdir -pv $PR_HOME
    ```

### Clone Repositories

1. Clone the main `data-fabric` repo, checking out the branch we'll use to test this PR.
    ```shell
    export DF_HOME=$PR_HOME/data-fabric
    git clone --branch dev https://github.com/raft-tech/data-fabric.git $DF_HOME
    ```

1. Clone the `df-arcadedb`, checking out the branch for this PR.
    ```shell
    export DF_ARCADEDB_HOME=$PR_HOME/df-arcadedb    
    git clone --branch YOUR_PR_BRANCH https://github.com/raft-tech/df-arcadedb.git $DF_ARCADEDB_HOME
    ```

1. Clone `dfdev` to help with building and running things. \
   ℹ️ _If you already have [`dfdev`](https://github.com/raft-tech/dfdev) you can skip this step._
    ```shell
    git clone https://github.com/raft-tech/dfdev.git $PR_HOME/dfdev
    PATH=$PATH:$PR_HOME/dfdev/dfdev
    ```
   
### Deploy

1. (Re)Create a new local `kind` cluster and deploy Data Fabric to it.
    ```shell
    dfdev cluster recreate
    ```

1. Re-build and re-deploy `df-arcadedb`

    ```shell
    dfdev redeploy df-arcadedb
    ``` 

## Verification Steps

- Set clearance to U and nationality to USA for the user in keycloak. Also make sure to leave the data steward group.
- Will fail since classification markings are missing

```CREATE VERTEX TYPE People```

- Next step will fail since overall classification is not set

```
INSERT INTO People CONTENT 
{
  firstName: 'Enzo',
  lastName: 'Ferrari'
}
```

- Will fail since classification markings are higher than what the user has access to

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo', 
  lastName: 'Ferrari', 
  classification: { clearance: 'TS', releasableTo: 'USA'}
}
```

- Will fail since classification markings are higher than what the user has access to

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo', 
  lastName: 'Ferrari', 
  classification: { clearance: 'S', releasableTo: 'USA'}
}
```

- Will fail since attributes are missing

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo', 
  lastName: 'Ferrari', 
  classification: { clearance: 'U', releasableTo: 'USA'}
}
```

- Will fail since classification markings are higher than what the user has access to

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo U',
  lastName: 'Ferrari',
  sources: [
    {
      name: 'IIR 123456',
      clearance: 'S',
      releasableTo: 'USA',
      attributes: {
        firstName: 'S',
        lastName: 'U'
      }
    }
  ]
}
```

- Will succeed

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo1', 
  lastName: 'Ferrari', 
  classification: { 
    clearance: 'U',
    releasableTo: 'USA',
    attributes: {
      firstName: 'U',
      lastName: 'U'
    }
  }
}
```

- Will succeed

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo2',
  lastName: 'Ferrari',
  sources: [
     {
       name: 'IIR 123456',
       clearance: 'U',
       releasableTo: 'USA',
       attributes: {
         firstName: 'U',
         lastName: 'U'
       }
     }
   ]
}
```

- In keycloak, bump the user's clearance attribute to S
- Will succeed

```
INSERT INTO People CONTENT 
{ 
 firstName: 'Enzo3', 
 lastName: 'Ferrari', 
  classification: { 
    clearance: 'S',
    releasableTo: 'USA',
    attributes: {
      firstName: 'S',
      lastName: 'U'
    }
  }
}
```

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo4',
  lastName: 'Ferrari',
  sources: [
     {
       name: 'IIR 123456',
       clearance: 'S',
       releasableTo: 'USA',
       attributes: {
         firstName: 'S',
         lastName: 'U'
       }
     }
   ]
}
```

```
INSERT INTO People CONTENT 
{ 
  firstName: 'Enzo5',
  lastName: 'Ferrari',
  sources: [
     {
       name: 'IIR 123456',
       clearance: 'S',
       releasableTo: 'USA',
       attributes: {
         firstName: 'S',
         lastName: 'U'
       }
     }
   ]
}
```

- select will render 5 objects

```select * from People```

- change nationality to GBR
- select will not render any data since all fields have `releasableTo` set to `USA`

```select * from People```

- change nationality back to USA and set clearance to U
- select will render 2 objects

```select * from People```

- Make sure to set clearance back to S
- demo relationship 

```CREATE VERTEX TYPE Location```

```
INSERT INTO Location CONTENT 
{ 
  name: 'Richmond',
  type: 'City',
  classification: {
    clearance: 'S',
    releasableTo: 'USA',
    attributes: {
      name: 'S',
      type: 'U'
    }
  }
}
```

```CREATE EDGE TYPE LivesIn```

- this will fail since classificatin markings are missing

```
CREATE EDGE LivesIn 
FROM (SELECT FROM People WHERE firstName = 'Enzo1') TO
     (SELECT FROM Location WHERE name = 'Richmond')
```

```
CREATE EDGE LivesIn 
FROM (SELECT FROM People WHERE firstName = 'Enzo1') TO
     (SELECT FROM Location WHERE name = 'Richmond')
CONTENT {
  sources: [
     {
       name: 'IIR 123456',
       clearance: 'S',
       releasableTo: 'USA',
       attributes: {
       }
     }
   ]
}
```

```SELECT FROM People WHERE firstName = 'Enzo1'```

- In the Arcade graph result GUI, select the returned node. In the context paragraph on the right side, select the both button to load incoming and outgoing edges. Confirm the LivesIn edge, and location Richmond load into the GUI.

- Finally show that switching nationality to GBR will not render previous relationship and vertex.
- change nationality to GBR. 
- this will filter out all data since we have set `releasableTo` to `USA`


```SELECT FROM People WHERE firstName = 'Enzo1'```
